### ISSUE TITLE
How to use for word2vec training?

### ISSUE THREAD

**Issue by Santosh-Gupta** [ddbourgin/numpy-ml#34]
Sorry for the newbie question, but I'm having a bit of trouble in trying to use the library. 

Specifically what I'm trying to do is embedding training, like word2vec. So I am trying to setup and embedding matrix, loss, and optimization using Adam. 

Any pointers would be greatly appreciated.

**Comment by ddbourgin** [comment]
Don't be sorry - this question is _very_ justified -- there is almost no usage documentation right now!

First, the bad news: at the moment, implementing word2vec will require a little bit of extra leg-work on your end. In particular, you'll need to implement either a negative sampler/noise contrastive estimation loss or a hierarchical softmax loss. For the latter, you could use the [`numpy_ml.preprocessing.nlp.HuffmanEncoder`](https://github.com/ddbourgin/numpy-ml/blob/fce2acfd7c370f55373bdc6dff1761a8258bfe27/numpy_ml/preprocessing/nlp.py#L409) module. 

Now, the good news: I'm actively working on writing an NCE loss object, and hope to push it ASAP. I'll also probably include a convenience `Embedding` layer to make embedding lookups a bit faster. I will update this thread when it has been pushed.

Ultimately, once these two components are in place, you should be able to write a relatively straightforward model. To see what a model object might look like, you can look at some examples in either the [`numpy_ml.neural_nets.modules`](https://github.com/ddbourgin/numpy-ml/blob/master/numpy_ml/neural_nets/modules/modules.py) or [`numpy_ml.neural_nets.models`](https://github.com/ddbourgin/numpy-ml/tree/master/numpy_ml/neural_nets/models) directories.

**Comment by ddbourgin** [comment]
Finally, one last caveat - if you're interested in training a non-toy word embedding model, I'd highly recommend using a library like keras, since it will make use of performance-optimized implementations of each model component. The code in this repo is meant to be clear and straightforward, but this often comes at the expense of efficiency!

**Comment by Santosh-Gupta** [comment]
Thanks, looking forward to it!

The reason why I am interested in doing it in Numpy is that Keras/Tensorflow isn't very great for sparse training. My use case is training 9 figures of embeddings. It eats up a lot of memory to have all those embeddings loaded into memory at the same time, and it's not necessary since only a small fraction is trained during each update step. 

So I was thinking of using your library and altering it so that it saves the unused weights to disk, and only loads them when they are trained or about to be trained.

**Comment by ddbourgin** [comment]
> So I was thinking of using your library and altering it so that it saves the unused weights to disk, and only loads them when they are trained or about to be trained.

That sounds like a good idea! If you end up implementing this, definitely consider submitting a PR :) I think this could be quite useful for a number of different model components, including the sparse evolutionary training layer (which currently uses dense matrices 😬). 

In the meantime, you might look into the [magnitude](https://github.com/plasticityai/magnitude) package (I haven't used it myself, but it seems potentially relevant).

**Comment by Santosh-Gupta** [comment]
It looks like this is exactly what I was looking for. I'm not familiar with a lot of computer science terms, but it uses sqlite as the datastore, so I'm guessing it does what I am looking for.

**Comment by Santosh-Gupta** [comment]
Looks like you can't use it for training, oh well 

https://github.com/plasticityai/magnitude/issues/32

**Comment by Santosh-Gupta** [comment]
But looking at the code gives me a strong idea of how I could implement it in this library, but using SQLite database as the vector store, and just copying the values back and forth between that and the numpy embedding array.

**Comment by ddbourgin** [comment]
Heya @Santosh-Gupta - I've just pushed a preliminary version of an `NCELoss` and word2vec model [here](https://github.com/ddbourgin/numpy-ml/blob/master/numpy_ml/neural_nets/models/w2v.py) and [here](https://github.com/ddbourgin/numpy-ml/blob/3bf1893083d3133f0fcd5fd4ce7624eeb4be03b3/numpy_ml/neural_nets/losses/losses.py#L464). 

Unfortunately, I suspect that if you're going to use the models for any sizeable dataset you'll have to do some performance modifications first. Let me know if you decide to try it out / have any questions in the meantime!

**Comment by Santosh-Gupta** [comment]
Sounds good!
